### PR TITLE
Support emitting function pointers in signatures

### PR DESCRIPTION
**Pull request by 333fred**
@AlekseyTs @dotnet/roslyn-compiler for review.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred It looks like tests are failing

File: src/Compilers/Core/Portable/PEWriter/Types.cs [hunk 1]
```diff
@@ -384,6 +384,17 @@ internal interface IPointerTypeReference : ITypeReference
         ITypeReference GetTargetType(EmitContext context);
     }
 
+    /// <summary>
+    /// This interface modesl the metadata representation of a pointer to a function in unmanaged memory.
```
Comment (cston) [inline src/Compilers/Core/Portable/PEWriter/Types.cs]: >modesl [](start = 23, length = 6)

Typo

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs [hunk 2]
```diff
@@ -81,21 +83,50 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
...
+
+            return new FunctionPointerMethodSymbol(
+                callingConvention,
+                refKind,
+                refCustomMods,
+                returnType,
+                retAndParamTypes.AsSpan()[1..]);
+        }
+
+        private static bool HasInModifier(ImmutableArray<CustomModifier> refCustomModifiers)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: >HasInModifier [](start = 28, length = 13)

Is this distinct from `CustomModifierUtils.HasInAttributeModifier()`?

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs [hunk 3]
```diff
@@ -110,6 +141,46 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
...
+
+                    paramsBuilder.Add(new FunctionPointerParameterSymbol(paramType, paramRefKind, i, this, paramRefCustomMods));
+                }
+
+                _parameters = paramsBuilder.ToImmutableAndFree();
+            }
+            else
+            {
+                _parameters = ImmutableArray<FunctionPointerParameterSymbol>.Empty;
+            }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: Consider extracting a `static ImmutableArray<ParameterSymbol> MakeParameters(...)` method.

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs [hunk 4]
```diff
@@ -131,6 +202,9 @@ internal bool Equals(FunctionPointerMethodSymbol other, TypeCompareKind compareK
         private bool EqualsNoParameters(FunctionPointerMethodSymbol other, TypeCompareKind compareKind, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverride)
             => CallingConvention == other.CallingConvention
                && RefKind == other.RefKind
+               && ((compareKind & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0
+                    ? RefCustomModifiers.SequenceEqual(other.RefCustomModifiers)
+                    : true)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: Perhaps:
```
((compareKind & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) != 0 ||
RefCustomModifiers.SequenceEqual(other.RefCustomModifiers))
```
Same comment for FunctionPointerParameterSymbol.

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
-    internal sealed class FunctionPointerParameterSymbol : ParameterSymbol
+    internal sealed partial class FunctionPointerParameterSymbol : ParameterSymbol
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs]: >partial [](start = 20, length = 7)

Is this change needed?

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs [hunk 6]
```diff
@@ -5,14 +5,16 @@
...
 using System.Reflection;
 using Microsoft.Cci;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
-    internal sealed class FunctionPointerMethodSymbol : MethodSymbol
+    internal sealed partial class FunctionPointerMethodSymbol : MethodSymbol
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: >partial [](start = 20, length = 7)

Is this needed?

File: src/Compilers/Core/Portable/PEWriter/Types.cs [hunk 7]
```diff
@@ -384,6 +384,17 @@ internal interface IPointerTypeReference : ITypeReference
...
+    /// </summary>
+    internal interface IFunctionPointerTypeReference : ITypeReference
+    {
+        /// <summary>
+        /// The signature of the function located at the target memory address.
+        /// </summary>
+        IMethodReference Signature { get; }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/Types.cs]: >IMethodReference [](start = 8, length = 16)

I think this should be ```ISignature```

File: src/Compilers/CSharp/Portable/Emitter/Model/FunctionPointerSymbolsAdapter.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+#nullable enable
+
+using System.Reflection.Metadata;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.Emit;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed partial class FunctionPointerTypeSymbol : Cci.IFunctionPointerTypeReference
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/FunctionPointerSymbolsAdapter.cs]: >FunctionPointerTypeSymbol [](start = 34, length = 25)

It looks like there is a typo in the name of the file - "FunctionPointerTypeSymbol**s**Adapter"

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs [hunk 9]
```diff
@@ -12,19 +12,21 @@
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
     // PROTOTYPE(func-ptr): Support generic substitution and retargeting
-    internal sealed class FunctionPointerTypeSymbol : TypeSymbol
+    // PROTOTYPE(func-ptr): Match msvc's emitting of custom modifiers for calling convention
+    internal sealed partial class FunctionPointerTypeSymbol : TypeSymbol
     {
         public static FunctionPointerTypeSymbol CreateFunctionPointerTypeSymbolFromSource(FunctionPointerTypeSyntax syntax, Binder typeBinder, DiagnosticBag diagnostics, ConsList<TypeSymbol> basesBeingResolved, bool suppressUseSiteDiagnostics)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs]: >FunctionPointerTypeSymbolFromSource [](start = 54, length = 35)

Consider shortening name to `Create` or `CreateFromSource`. Similar comment for `Create...FromMetadata` below.

File: src/Compilers/Core/Portable/PEWriter/MetadataVisitor.cs [hunk 10]
```diff
@@ -465,6 +465,17 @@ public virtual void Visit(IPointerTypeReference pointerTypeReference)
             this.Visit(pointerTypeReference.GetTargetType(Context));
         }
 
+        public virtual void Visit(IFunctionPointerTypeReference functionPointerTypeReference)
+        {
+            this.Visit(functionPointerTypeReference.Signature.RefCustomModifiers);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/MetadataVisitor.cs]: >RefCustomModifiers [](start = 62, length = 18)

What about ReturnValueCustomModifiers?

File: src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs [hunk 11]
```diff
@@ -758,5 +647,48 @@ class C
...
+                foreach (var param2 in m.Parameters)
+                {
+                    if (!ReferenceEquals(param1, param2))
+                    {
+                        Assert.False(param1.Type.Equals(param2.Type, TypeCompareKind.ConsiderEverything));
+                    }
+                    else
+                    {
+                        Assert.True(param1.Type.Equals(param2.Type, TypeCompareKind.ConsiderEverything));
+                    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs]:
```
Assert.NotEqual(
    ReferenceEquals(param1, param2),
    param1.Type.Equals(param2.Type, TypeCompareKind.ConsiderEverything));
```

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 12]
```diff
@@ -147,6 +147,26 @@ public override void VisitPointerType(IPointerTypeSymbol symbol)
             AddPunctuation(SyntaxKind.AsteriskToken);
         }
 
+        public override void VisitFunctionPointerType(IFunctionPointerTypeSymbol symbol)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs]: >VisitFunctionPointerType [](start = 29, length = 24)

Are we testing this?

File: src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs [hunk 13]
```diff
@@ -4,6 +4,7 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using Microsoft.Cci;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs]: I think this namespace was not imported intentionally.

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: Only in that I did not realize the latter existed :)

---
In reply to: [374321386](https://github.com/dotnet/roslyn/pull/41340#discussion_r374321386) [](ancestors = 374321386)

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs]: Not anymore, thought I was going to need to create a specific Cci adapter.

---
In reply to: [374336827](https://github.com/dotnet/roslyn/pull/41340#discussion_r374336827) [](ancestors = 374336827)

File: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs [hunk 14]
```diff
@@ -319,6 +319,24 @@ private TypeSymbol DecodeTypeOrThrow(ref BlobReader ppSig, SignatureTypeCode typ
                     typeSymbol = DecodeGenericTypeInstanceOrThrow(ref ppSig, out refersToNoPiaLocalType);
                     break;
 
+                case SignatureTypeCode.FunctionPointer:
+                    var signatureHeader = ppSig.ReadSignatureHeader();
+                    var parameters = DecodeSignatureParametersOrThrow(ref ppSig, signatureHeader, typeParameterCount: out int typeParamCount, shouldProcessAllBytes: false);
+                    Debug.Assert(typeParamCount == 0);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: >Debug.Assert(typeParamCount == 0); [](start = 20, length = 34)

I think we should do a check and ```throw new UnsupportedSignatureContent()``` otherwise.

File: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs [hunk 15]
```diff
@@ -319,6 +319,24 @@ private TypeSymbol DecodeTypeOrThrow(ref BlobReader ppSig, SignatureTypeCode typ
...
+                    Cci.CallingConvention callingConvention = signatureHeader.CallingConvention switch
+                    {
+                        SignatureCallingConvention.Default => Cci.CallingConvention.Default,
+                        SignatureCallingConvention.CDecl => Cci.CallingConvention.CDecl,
+                        SignatureCallingConvention.FastCall => Cci.CallingConvention.FastCall,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: >Cci.CallingConvention.FastCall [](start = 63, length = 30)

It looks like we will not be able to serialize a type like that. Is there a reason for that? In general, we should be able to gracefully handle everything we import, but it looks like MetadatWriter is going to crash on this convention.

File: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs [hunk 16]
```diff
@@ -319,6 +319,24 @@ private TypeSymbol DecodeTypeOrThrow(ref BlobReader ppSig, SignatureTypeCode typ
...
+                        SignatureCallingConvention.StdCall => Cci.CallingConvention.Standard,
+                        SignatureCallingConvention.ThisCall => Cci.CallingConvention.ThisCall,
+                        SignatureCallingConvention.VarArgs => Cci.CallingConvention.ExtraArguments,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: >Cci.CallingConvention.ExtraArguments [](start = 62, length = 36)

Same comment

File: src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs (hunk 11 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs]: I think the way it's written currently is easier to understand and diagnose if something went wrong. I'm going to leave it as is.

---
In reply to: [374345105](https://github.com/dotnet/roslyn/pull/41340#discussion_r374345105) [](ancestors = 374345105)

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs (hunk 12 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs]: No, I just added it for ease of debugging (GetDebuggerDisplay uses it). I'll add a prototype, as the formatting is kinda wonky and I'm not planning on addressing this until I address the public API in general.

---
In reply to: [374345748](https://github.com/dotnet/roslyn/pull/41340#discussion_r374345748) [](ancestors = 374345748)

File: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs (hunk 15 above)
Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: As currently spec'd, C# function pointers don't have fastcall. I'll raise the question and make this throw for now.

---
In reply to: [374388482](https://github.com/dotnet/roslyn/pull/41340#discussion_r374388482) [](ancestors = 374388482)

File: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs (hunk 16 above)
Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: This one should probably just throw, I don't think we're actually planning on supporting varags function pointers in C#.

---
In reply to: [374388711](https://github.com/dotnet/roslyn/pull/41340#discussion_r374388711) [](ancestors = 374388711)

File: src/Compilers/CSharp/Portable/Emitter/Model/FunctionPointerSymbolsAdapter.cs [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        IMethodReference IFunctionPointerTypeReference.Signature => Signature;
+        void IReference.Dispatch(Cci.MetadataVisitor visitor) => visitor.Visit((IFunctionPointerTypeReference)this);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/FunctionPointerSymbolsAdapter.cs]: >IReference.Dispatch [](start = 13, length = 19)

It feels like we should also implement two other APIs of IReference

File: src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs (hunk 13 above)
Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs]: It is, `CallingConvention` is a Cci type. If you prefer I can qualify that one instance and remove the import.

---
In reply to: [374347237](https://github.com/dotnet/roslyn/pull/41340#discussion_r374347237) [](ancestors = 374347237)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs [hunk 18]
```diff
@@ -162,9 +162,11 @@ private TypeSymbol DecodeType(TypeSymbol type)
                         _foundUsableErrorType = true;
                     }
                     return type;
+
                 case SymbolKind.DynamicType:
                 case SymbolKind.TypeParameter:
                 case SymbolKind.PointerType:
+                case SymbolKind.FunctionPointer:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs]: >case SymbolKind.FunctionPointer: [](start = 16, length = 32)

This doesn't feel right.

**Comment by AlekseyTs** [comment]
>     internal struct TupleTypeDecoder

It feels like other decoders should also be adjusted: nullable, dynamic, etc.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs:63 in c83dd64. [](commit_id = c83dd642d5f722a90583a2058438a16a8fc562ea, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/PublicModel/FunctionPointerTypeSymbol.cs [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel
+{
+    internal sealed class FunctionPointerTypeSymbol : TypeSymbol, IFunctionPointerTypeSymbol
+    {
+        private readonly Symbols.FunctionPointerTypeSymbol _underlying;
+        private IMethodSymbol? _lazySignatureSymbol;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/PublicModel/FunctionPointerTypeSymbol.cs]: >private IMethodSymbol? _lazySignatureSymbol; [](start = 8, length = 44)

I am not sure if this cache really needed. I think each symbol already caches its public representation

File: src/Compilers/CSharp/Portable/Symbols/PublicModel/FunctionPointerTypeSymbol.cs [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+        [return: MaybeNull]
+        protected override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
+        {
+            return visitor.VisitFunctionPointerType(this);
+        }
+
+        protected override ITypeSymbol WithNullableAnnotation(CodeAnalysis.NullableAnnotation nullableAnnotation)
+        {
+            Debug.Assert(nullableAnnotation != this.NullableAnnotation);
+            return new FunctionPointerTypeSymbol(_underlying, nullableAnnotation);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/PublicModel/FunctionPointerTypeSymbol.cs]: >return [](start = 12, length = 6)

```Debug.Assert(nullableAnnotation != _underlying.DefaultNullableAnnotation);```?

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs [hunk 21]
```diff
@@ -83,24 +84,22 @@ internal override bool Equals(TypeSymbol t2, TypeCompareKind compareKind, IReadO
...
         {
-            return Hash.Combine(1, _signature.GetHashCode());
+            return Hash.Combine(1, Signature.GetHashCode());
         }
 
         protected override ISymbol CreateISymbol()
         {
-            // PROTOTYPE(func-ptr): Implement
-            throw new NotImplementedException();
+            return CreateITypeSymbol(CodeAnalysis.NullableAnnotation.None);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs]: >CodeAnalysis.NullableAnnotation.None [](start = 37, length = 36)

DefaultNullableAnnotation?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs]: >return CreateITypeSymbol [](start = 12, length = 24)

I think the code reuse doesn't give us a lot of benefits, but disallows us to assert that we properly use internal cash for public symbols.

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs [hunk 22]
```diff
@@ -83,24 +84,22 @@ internal override bool Equals(TypeSymbol t2, TypeCompareKind compareKind, IReadO
...
         }
 
         protected override ITypeSymbol CreateITypeSymbol(CodeAnalysis.NullableAnnotation nullableAnnotation)
         {
-            // PROTOTYPE(func-ptr): Implement
-            throw new NotImplementedException();
+            return new PublicModel.FunctionPointerTypeSymbol(this, nullableAnnotation);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs]: >return [](start = 12, length = 6)

```Debug.Assert(nullableAnnotation != DefaultNullableAnnotation);```?

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs [hunk 23]
```diff
@@ -81,21 +83,50 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
...
                 typeBinder,
                 diagnostics,
                 suppressUseSiteDiagnostics);
         }
 
+        public static FunctionPointerMethodSymbol CreateMethodFromMetadata(CallingConvention callingConvention, ImmutableArray<ParamInfo<TypeSymbol>> retAndParamTypes)
+        {
+            Debug.Assert(retAndParamTypes.Length > 0);
+
+            ParamInfo<TypeSymbol> retInfo = retAndParamTypes[0];
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: >ParamInfo<TypeSymbol> retInfo = retAndParamTypes[0] [](start = 12, length = 51)

Why can't we do all this return type related work within the constructor?

File: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs (hunk 16 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: > This one should probably just throw, I don't think we're actually planning on supporting varags function pointers in C#.

What we should be able to import/emit and what you can define in source code is not necessary the same thing. I think that up until the language is getting involved and symbol model is able to represent the metadata, we shouldn't fail

---
In reply to: [374392051](https://github.com/dotnet/roslyn/pull/41340#discussion_r374392051) [](ancestors = 374392051,374388711)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs [hunk 24]
```diff
@@ -0,0 +1,342 @@
...
+using System;
+using System.Collections.Immutable;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
+{
+    public class CodeGenFunctionPointersTests : CSharpTestBase
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs]: >CodeGenFunctionPointersTests [](start = 17, length = 28)

It might be useful to have targeted tests for metadata import from IL to ensure that things end up in the right places.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs (hunk 13 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs]: > If you prefer I can qualify that one instance and remove the import.

Yes, I would prefer we do this as we are doing for other Cci type references in this file.

---
In reply to: [374393583](https://github.com/dotnet/roslyn/pull/41340#discussion_r374393583) [](ancestors = 374393583,374347237)

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs [hunk 25]
```diff
@@ -3695,6 +3701,24 @@ private void SerializeTypeReference(SignatureTypeEncoder encoder, ITypeReference
...
                 }
 
+                if (typeReference is IFunctionPointerTypeReference functionPointerTypeReference)
+                {
+                    var signature = functionPointerTypeReference.Signature;
+                    var signatureEncoder = encoder.FunctionPointer(convention: convertCallingConvention(signature.CallingConvention));
+                    SerializeReturnValueAndParameters(signatureEncoder, signature, varargParameters: ImmutableArray<IParameterTypeInformation>.Empty);
+                    return;
+
+                    static SignatureCallingConvention convertCallingConvention(CallingConvention originalConvention)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs]: >SignatureCallingConvention [](start = 27, length = 26)

It feels like translating between two enums is just another hoop to jump over. Would it be possible to remove the Cci version of the enum and switch to using ```SignatureCallingConvention```, which, I assume, is a public type and could be used in our public APIs as well?

**Comment by 333fred** [comment]
>     internal struct TupleTypeDecoder

I'll put a prototype. I was specifically avoiding them in this pr.

---
In reply to: [581664963](https://github.com/dotnet/roslyn/pull/41340#issuecomment-581664963) [](ancestors = 581664963)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs:63 in c83dd64. [](commit_id = c83dd642d5f722a90583a2058438a16a8fc562ea, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs (hunk 23 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: We could. But given that Chuck just asked me to move what _was_ in the constructor into a helper method, I'm going to leave this where it is.

---
In reply to: [374405675](https://github.com/dotnet/roslyn/pull/41340#discussion_r374405675) [](ancestors = 374405675)

**Comment by 333fred** [comment]
@AlekseyTs @cston addressed feedback.

---
In reply to: [581682378](https://github.com/dotnet/roslyn/pull/41340#issuecomment-581682378) [](ancestors = 581682378)

File: src/Compilers/Core/Portable/PEWriter/Members.cs [hunk 26]
```diff
@@ -70,6 +70,23 @@ internal enum CallingConvention
...
+        private const SignatureCallingConvention SignatureMask =
+            SignatureCallingConvention.Default
+            | SignatureCallingConvention.CDecl
+            | SignatureCallingConvention.StdCall
+            | SignatureCallingConvention.ThisCall
+            | SignatureCallingConvention.FastCall
+            | SignatureCallingConvention.VarArgs;
+
+        internal static CallingConvention FromSignatureConvention(this SignatureCallingConvention convention)
+            => (CallingConvention)(convention & SignatureMask);
```
Comment (cston) [inline src/Compilers/Core/Portable/PEWriter/Members.cs]: >convention & SignatureMask [](start = 35, length = 26)

Are we silently ignoring bits outside `SignatureMask`? Should we `throw UnsupportedSignatureContent()` in those cases instead?

File: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs [hunk 27]
```diff
@@ -319,6 +319,18 @@ private TypeSymbol DecodeTypeOrThrow(ref BlobReader ppSig, SignatureTypeCode typ
...
+                case SignatureTypeCode.FunctionPointer:
+                    var signatureHeader = ppSig.ReadSignatureHeader();
+                    var parameters = DecodeSignatureParametersOrThrow(ref ppSig, signatureHeader, typeParameterCount: out int typeParamCount, shouldProcessAllBytes: false);
+
+                    if (typeParamCount != 0)
+                    {
+                        throw new UnsupportedSignatureContent();
+                    }
+
+                    typeSymbol = MakeFunctionPointerTypeSymbol(Cci.CallingConventionUtils.FromSignatureConvention(signatureHeader.CallingConvention), ImmutableArray.Create(parameters));
```
Comment (cston) [inline src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: >Cci.CallingConventionUtils.FromSignatureConvention(signatureHeader.CallingConvention) [](start = 63, length = 85)

`signatureHeader.CallingConvention.FromSignatureConvention()`

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs [hunk 28]
```diff
@@ -81,21 +83,77 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
...
+                    ParamInfo<TypeSymbol> param = parameterTypes[i];
+                    var paramRefCustomMods = CSharpCustomModifier.Convert(param.RefCustomModifiers);
+                    var paramType = TypeWithAnnotations.Create(param.Type, customModifiers: CSharpCustomModifier.Convert(param.CustomModifiers));
+                    // PROTOTYPE(func-ptr): Need to encode out params as a custom modifier of some kind
+                    var paramRefKind = param.IsByRef switch
+                    {
+                        false => RefKind.None,
+                        true when CustomModifierUtils.HasInAttributeModifier(paramRefCustomMods) => RefKind.In,
+                        true => RefKind.Ref,
+                    };
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: Consider extracting a helper method for use here and in `CreateFromMetadata()` above.

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs [hunk 29]
```diff
@@ -81,21 +83,77 @@ public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointer
...
+
+            return new FunctionPointerMethodSymbol(
+                callingConvention,
+                refKind,
+                refCustomMods,
+                returnType,
+                retAndParamTypes.AsSpan()[1..]);
+        }
+
+        private static ImmutableArray<FunctionPointerParameterSymbol> MakeParametersFromMetadata(ReadOnlySpan<ParamInfo<TypeSymbol>> parameterTypes, FunctionPointerMethodSymbol parent)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: >MakeParametersFromMetadata [](start = 70, length = 26)

It looks like there is only one call site, consider making this a local function

File: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs (hunk 23 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs]: > We could. But given that Chuck just asked me to move what was in the constructor into a helper method, I'm going to leave this where it is.

I am not sure how one is related to the other. I think, from the API design perspective, it would be better if constructor were taking all data in raw form, or all data in "preprocessed" form.


---
In reply to: [374952548](https://github.com/dotnet/roslyn/pull/41340#discussion_r374952548) [](ancestors = 374952548,374405675)

File: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs (hunk 27 above)
Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: Actually, there aren't that many here, I'll just do it.

---
In reply to: [375578610](https://github.com/dotnet/roslyn/pull/41340#discussion_r375578610) [](ancestors = 375578610,375063146)

Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs]: > there aren't that many here

But there _are_ conflicting names between Cci. and SRM., so I'll leave it.

---
In reply to: [375582556](https://github.com/dotnet/roslyn/pull/41340#discussion_r375582556) [](ancestors = 375582556,375578610,375063146)
